### PR TITLE
repr,pgrepr: tweak list encoding and decoding

### PR DESCRIPTION
**Pull request by benesch**
See commit messages for details.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/2776)

### CODE REVIEW

**Comment by benesch** [comment]
There are roughly two chunks of this. Chunk 1 removes some clones by sharing a buffer. Chunk 2 builds on chunk 1 and removes all extra allocations, but requires a tiny bit of unsafe code.

Performance improvement from master to chunk 1:

```
format_list simple      time:   [185.68 us 187.90 us 190.38 us]
                        change: [-64.141% -63.722% -63.274%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

format_list nested      time:   [114.62 us 115.54 us 116.54 us]
                        change: [-81.913% -81.757% -81.605%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe
```

Performance improvement from chunk 1 to chunk 2:

```
Gnuplot not found, using plotters backend
format_list simple      time:   [158.71 us 159.36 us 160.08 us]                               
                        change: [-17.354% -16.369% -15.456%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe

format_list nested      time:   [27.192 us 27.294 us 27.408 us]                                
                        change: [-76.335% -76.133% -75.915%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe
```

Performance improvement from master to chunk 2:

```
format_list simple      time:   [165.31 us 167.10 us 168.87 us]                               
                        change: [-68.941% -68.522% -67.966%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

format_list nested      time:   [23.223 us 23.388 us 23.578 us]                                
                        change: [-96.309% -96.281% -96.252%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe
```

File: test/sqllogictest/list.slt [hunk 1]
```diff
@@ -56,4 +56,9 @@ select LIST['foo', 'f}o', '"\', null, 'null', 'NULL']
 query T
 select list[list[list['"']]];
 ----
-{"{\"{\\\"\\\\\\\"\\\"}\"}"}
+{{{"\""}}}
```
Comment (jamii) [inline test/sqllogictest/list.slt:59]: What does `select list['{1}']` print?

**Comment by jamii** [comment]
Code looks reasonable.

We're eating a lot of complexity to handle formatting into both strings and bytesmut. What if we just always format into string and eat a single copy per datum in pgwire?

File: test/sqllogictest/list.slt (hunk 1 above)
Comment (benesch) [reply test/sqllogictest/list.slt:59]:
```
{"{1}"}
```

(added as a test case)

File: test/lang/js/params.test.ts [hunk 2]
```diff
@@ -52,4 +52,38 @@ describe("query api", () => {
     });
     expect(res.rows).toEqual([["12"]]);
   });
+
+  describe("list parameters", () => {
+    it("should handle a simple int array", async () => {
+      const res = await client.query({
```
[the remark this answers is not in the document]
Comment (benesch) [reply test/lang/js/params.test.ts:58]: I mean, that _is_ what parameter decoding is. :D But yeah, this is actually sending the query and the parameters separately over the wire, not mashing strings together or anything.

**Comment by benesch** [comment]
> We're eating a lot of complexity to handle formatting into both strings and bytesmut. What if we just always format into string and eat a single copy per datum in pgwire?

Do you think it would make things that much simpler? You'd be able to get rid of everything in `ore::fmt`, but that's a fairly localized and straightforward re-exportation of some common bits of `String` and `BytesMut`. Otherwise it's just a bunch of signatures like

```rust
pub fn format_bytes<F>(buf: &mut F, bytes: &[u8]) -> Nestable
where
    F: FormatBuffer;
```

that would become:

```rust
pub fn format_bytes(buf: &mut String, bytes: &[u8]) -> Nestable;
```

I guess maybe nothing seems bad in comparison to differential. :D

**Comment by jamii** [comment]
Oops, sorry I missed this.

I'm not opposed to merging, just wanted to float the idea.

**Comment by benesch** [comment]
👍 I've filed an issue (#2952) that will remind me to re-evaluate when we have some more experience, especially when we have record types.

**Pull request merged**